CI tasks (code analysis, static security analysis, test and coverage workflows)

asep-fajar-firmansyah opened the pull request:
Implement linting and slither analyzer for a static security audit

asep-fajar-firmansyah committed:
modified:   contracts/BnomialNFT.sol
	modified:   package-lock.json
	modified:   package.json

asep-fajar-firmansyah committed:
new file:   .solhint.json
	new file:   .solhintignore
	new file:   linter-log.txt
	new file:   tslint.json

asep-fajar-firmansyah committed:
modified:   .solhint.json

asep-fajar-firmansyah committed:
added a static security audit

asep-fajar-firmansyah committed:
deleted:    linter-log.txt

asep-fajar-firmansyah committed:
added linter log

asep-fajar-firmansyah committed:
modified:   .solhint.json

haltakov commented on contracts/BnomialNFT.sol:
```diff
@@ -62,17 +62,17 @@ contract BnomialNFT is ERC721, Ownable {
         }
 
-        string memory part0 = '{"name":"Bnomial Badges",';
+        string memory part0 = "{\"name\":\"Bnomial Badges\",";
```
I guess these changes come from some automatic formatting. Do you use `prettier`? I think we should add a standard config.

asep-fajar-firmansyah replied:
No, I didn't use it. Okay, I will apply prettier.

haltakov commented on linter-log.txt:
```diff
@@ -0,0 +1,29 @@
+
```
We should have the log files ignored in .gitignore and not check them in

haltakov commented on package.json:
```diff
@@ -10,23 +10,28 @@
         "verify:polygon": "npx hardhat verify --network polygon",
-        "verify:mumbai": "npx hardhat verify --network mumbai"
+        "verify:mumbai": "npx hardhat verify --network mumbai",
+        "lint": "solhint --max-warnings 0 \"contracts/**/*.sol\""
```
I like that! :)

haltakov commented on slither-audit.txt:
```diff
@@ -0,0 +1,62 @@
+
```
We should have the log files ignored in .gitignore and not check them in

asep-fajar-firmansyah replied:
Ahh okay, I'll implement it

haltakov commented on tslint.json:
```diff
@@ -0,0 +1,16 @@
+{
```
Hmm, why do we need a lint config for TypeScript?

asep-fajar-firmansyah replied:
Actually, I didn't use it. I'll delete it

haltakov commented:
Thanks for submitting the PR. I like the changes, but there are small things to fix :)

kfischer-okarin reviewed (changes requested):
Maybe I'm missing something but where is the actual CI? There should be a `.github/workflows/tests.yml` (or whatever you name the yaml file) with a job actually running all that stuff...

And I agree with the comments so far that the log files should be added to `.gitignore` and deleted from this branch

haltakov commented:
Yes @kfischer-okarin, the CI config is still missing. I don't mind having separate PRs for the lint and the actual CI :)

asep-fajar-firmansyah commented:
Thanks @kfischer-okarin for the hint, I'll investigate and implement it.

kfischer-okarin commented:
Well then I guess it's up to @asep-fajar-firmansyah ...
- Either just remove the log files, add them to gitignore and maybe rename this Pull Request to something along the lines of "add linter configuration"
- Or also add the CI config in the same Pull Request

Btw if you need an example:
Here's the CI configuration I implemented for the API repository (though in the end we didn't end up using it)

https://github.com/underfitted/bnomial/pull/25/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957f

In your case you need to use https://github.com/actions/setup-node instead of setup-python.... but the other parts should be easily adaptable I think

asep-fajar-firmansyah committed:
add config and implement prettier

asep-fajar-firmansyah committed:
deleted:    .solhintignore

asep-fajar-firmansyah committed:
new file:   lint.yml

asep-fajar-firmansyah committed:
modefied lint config

asep-fajar-firmansyah committed:
modified lint config

asep-fajar-firmansyah committed:
modified lint config

asep-fajar-firmansyah committed:
modified lint config

asep-fajar-firmansyah committed:
modified lint config

asep-fajar-firmansyah committed:
Testing CI pipeline using solhint

asep-fajar-firmansyah committed:
modified:   package.json

asep-fajar-firmansyah committed:
testing continues on error-true

asep-fajar-firmansyah committed:
update lint config

asep-fajar-firmansyah committed:
testing ci using lint and slither

asep-fajar-firmansyah committed:
testing ci using lint and slither

asep-fajar-firmansyah committed:
modified:   .gitignore

asep-fajar-firmansyah committed:
testing ci using lint and slither

asep-fajar-firmansyah commented:
Hi @haltakov, I added linter and slither-analyzer configuration as parts of CI to do code analysis and static analysis, respectively. And please review as well.

asep-fajar-firmansyah committed:
update .gitignore

haltakov commented:
> Hi @haltakov, I added linter and slither-analyzer configuration as parts of CI to do code analysis and static analysis, respectively. And please review as well.

Thanks for the changes! I'm not very experienced with CI, so I have to look into this in detail... In the meantime - we still need to add the tests and the coverage to the CI, right?

asep-fajar-firmansyah commented:
Aah okay, I will add test and the coverage.

asep-fajar-firmansyah committed:
added test and coverage workflow

asep-fajar-firmansyah commented:
I added test and coverage workflows in test_and_coverage.yml.

asep-fajar-firmansyah committed:
testing CI

kfischer-okarin commented on package.json:
```diff
@@ -10,23 +10,29 @@
+        "@nomiclabs/hardhat-ethers": "^2.0.3",
+        "@nomiclabs/hardhat-etherscan": "^2.1.8",
         "@nomiclabs/hardhat-waffle": "^2.0.1",
+        "@openzeppelin/contracts": "^4.4.1",
```
I'm not quite sure about this being moved to devDependencies - since this contains actual Solidity building blocks used by the application itself right? So this should still be in dependencies...

asep-fajar-firmansyah replied:
I removed unused libraries

kfischer-okarin commented on package.json:
```diff
@@ -10,23 +10,29 @@
+        "prettier": "2.5.1",
+        "solhint": "^3.3.6",
+        "solidity-coverage": "^0.7.17",
+        "typescript": "^4.5.4"
```
Is this project using typescript?

asep-fajar-firmansyah replied:
I removed unused libraries

kfischer-okarin commented on package.json:
```diff
@@ -10,23 +10,29 @@
     "dependencies": {
-        "@openzeppelin/contracts": "^4.3.3",
-        "dotenv": "^10.0.0"
+        "ci": "^2.1.1"
```
Is this library actually used anywhere?

asep-fajar-firmansyah replied:
I removed unused libraries

kfischer-okarin commented on .env:
```diff
@@ -0,0 +1,4 @@
+MUMBAI_API_URL=""
+POLYGON_API_URL=""
+PRIVATE_KEY="0000000000000000000000000000000000000000000000000000000000000000"
+POLYSCAN_KEY=""
```
`.env` are not supposed to be committed they are gitignored files to be used by each developer to put their access keys inside so they can be accessed during development....

This file mainly seems to serve the purpose of putting dummy values for testing inside... I don't think it's needed....

A better approach for this problem would be write a wrapper scripts for running the tests....something like

```bash
#!/bin/bash

# Dummy values for test
export MUMBAI_API_URL=""
export POLYGON_API_URL=""
export PRIVATE_KEY="0000000000000000000000000000000000000000000000000000000000000000"
export POLYSCAN_KEY=""

hardhat test
```

Putting these dummy values in a wrapper script would have the advantage of still actively requiring them for actual deployment

asep-fajar-firmansyah replied:
I added it into the workflow and deleted .env file

kfischer-okarin commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,23 @@
+        # To load bashrc
+        shell: bash -ieo pipefail {0}
+
+on: [push, pull_request]
```
```yml
on:
  push:
    branches:
      - main
  pull_request:
```

Is probably the better setting ? Currently it will run every workflow twice on pull request branches... once for the push and once for the update of the pull request triggered by the push.... With the settings above it would only run once for pull request pushes/updates and once when it's actually merged....

Same for the other two workflows I guess

asep-fajar-firmansyah replied:
I got it, I'll implement it

kfischer-okarin commented on .github/workflows/slither.yml:
```diff
@@ -0,0 +1,59 @@
+        paths:
+            - 'contracts/**'
+            - '.github/workflows/slither.yml'
+            - '**.py'
```
Are there any python files in this project?

asep-fajar-firmansyah replied:
No, there are no python files. So, I deleted the code of this line

kfischer-okarin commented on .github/workflows/slither.yml:
```diff
@@ -0,0 +1,59 @@
+              run: |
+                  echo "Running slither"
+                  slither . --print human-summary
+            - name: Contract summary of static ananlysis
+              run: |
+                  echo "Running slither"
+                  slither . --print contract-summary
```
I'm not familiar with all this stuff - but is there something more useful than just printing all kinds of information? Are there like warning thresholds to employ?

For example this step ("Summary of static analysis") in:
https://github.com/underfitted/bnomial-web3/runs/4622459514

I see a "high issues" count of 2... is that ok? Should the step maybe throw an error then?

asep-fajar-firmansyah replied:
I changed the workflow. So, I just reported the issues to do solved

kfischer-okarin commented on .github/workflows/slither.yml:
```diff
@@ -0,0 +1,59 @@
+                  slither . --print function-summary
+            - name: Data dependency
+              run: |
+                  slither . --print data-dependency
```
Same concern as above... especially this step just seems to print thousands of lines of data dependency.... is anyone going to read that? Anything more useful to be done here?

asep-fajar-firmansyah replied:
I changed the workflow. So, I just reported the issues to do solved

asep-fajar-firmansyah committed:
delete ci workflows

asep-fajar-firmansyah committed:
delete .env file

asep-fajar-firmansyah committed:
update package.json

asep-fajar-firmansyah committed:
update package.json

asep-fajar-firmansyah committed:
applying CI

asep-fajar-firmansyah committed:
deleted unused file

asep-fajar-firmansyah committed:
testing CI on push mode

asep-fajar-firmansyah committed:
testing CI on push mode

asep-fajar-firmansyah committed:
Updated CI files

asep-fajar-firmansyah committed:
Testing static security audit using slither

asep-fajar-firmansyah commented:
Hi @haltakov, @kfischer-okarin 
I have updated the CI workflows, including test and coverage, code analysis using solhint, and static security audit using slither-analyzer. And please review as well.

More detail:
1. The code analysis using solhint. I use the following configuration:
```
{
    "extends": "solhint:recommended",
    "rules": {
        "max-line-length": "off",
        "avoid-suicide": "error",
        "avoid-sha3": "warn",
        "compiler-version": ["error", "^0.8.2"],
        "func-visibility": ["warn", { "ignoreConstructors": true }],
        "no-empty-blocks": "off",
        "reason-string": ["off"],
        "no-inline-assembly": "off",
        "quotes": "off"
    }
}
```
I used recommended mode in the config refers to this reference (https://protofire.github.io/solhint/). However, I modified the configuration to avoid errors, as attached below if we use the suggestion configuration according to the reference.
![image](https://user-images.githubusercontent.com/42900622/147611886-8197302b-422b-4b01-91db-010f77e732a0.png)

Would you please review the solhint configuration to determine whether the configuration has satisfied our needs or not? 

2. The static security audit using slither
As shown in the link below, I found some issues to solve that show the CI workflow is working. 
https://github.com/underfitted/bnomial-web3/actions/runs/1631965602
You can download the report and have a look at the issues.

asep-fajar-firmansyah committed:
updated package.json

asep-fajar-firmansyah committed:
deleted:    scripts/deploy.js

asep-fajar-firmansyah committed:
new file:   scripts/deployNFT.js
	new file:   scripts/deployToken.js

asep-fajar-firmansyah committed:
modified:   test/BnomialNFT.test.js

asep-fajar-firmansyah committed:
new file:   test/BnomialToken.test.js

asep-fajar-firmansyah committed:
updated package.json

asep-fajar-firmansyah committed:
Merge branch 'main' into setup_ci_for_sc_verification

asep-fajar-firmansyah committed:
bugs fixing based on the static audit suggestions

asep-fajar-firmansyah commented:
Hi @haltakov, I have updated the repo with some bug fixing based on static audit suggestions, which are solved 12 of 21 issues. And then, I increased the FALSE_POSITIVE_WARNINGS value from 7 to 9, which means if the slither-analyzer identifies the warning less than and equal nine, then the system will be notified passed checking. 

The previous audit results can be accessed at https://github.com/underfitted/bnomial-web3/actions/runs/1632170903

asep-fajar-firmansyah committed:
exclude-informational analysis results on static audit

asep-fajar-firmansyah committed:
Set FALSE_POSITIVE_WARNINGS=5 on slither config

asep-fajar-firmansyah commented:
Hi @haltakov, I updated slither.yml by adding --exclude-informational to reduce issues inform of information impact analysis. So, there are 3 issues that remain, as follows:
1. Base64.encode(bytes) (contracts/Base64.sol#12-71) contains an incorrect shift operation: mstore(uint256,uint256)(resultPtr_encode_asm_0 - 2,0x3d3d << 240) (contracts/Base64.sol#61)
2. Base64.encode(bytes) (contracts/Base64.sol#12-71) contains an incorrect shift operation: mstore(uint256,uint256)(resultPtr_encode_asm_0 - 1,0x3d << 248) (contracts/Base64.sol#64)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#shift-parameter-mixup
3. tokenURI(uint256) should be declared external:
	- BnomialNFT.tokenURI(uint256) (contracts/BnomialNFT.sol#48-101)
	- ERC721.tokenURI(uint256) (node_modules/@openzeppelin/contracts/token/ERC721/ERC721.sol#93-98)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#public-function-that-could-be-declared-external.
For issue number 3, I need your help to solve it.

This is a static audit report: https://github.com/underfitted/bnomial-web3/actions/runs/1638885744

As long as BnomialNFT.sol uses Base64.sol, the slither-analyzer will always analyze this file. However, we can accommodate the issues from the analysis results from Base64.sol by setting FALSE_POSITIVE_WARNINGS that determine how many issues we can ignore. In addition, I set FALSE_POSITIVE_WARNINGS to 5 to reduce the risk because of errors/warnings.

kfischer-okarin commented on .github/workflows/slither.yml:
```diff
@@ -0,0 +1,73 @@
+        run: |
+          echo "Running slither"
+          #slither . --ignore-compile --filter-paths "node_modules" --disable-color --print human-summary >>.github/slither-output/slither_report.txt 2>&1 || true
+          slither . --json-types compilations,console,detectors,printers,list-detectors,list-printers --exclude-dependencies --exclude-informational --compile-force-framework hardhat >>.github/slither-output/slither_report.txt 2>&1 || true
```
```suggestion
          slither . --json-types compilations,console,detectors,printers,list-detectors,list-printers --exclude-dependencies --exclude-informational --compile-force-framework hardhat 2>&1 | tee -a .github/slither-output/slither_report.txt
```

should do the trick and also print it out to stdout at the same time.... (tee always succeeds)

kfischer-okarin replied:
But then again.... Do we need this report as a downloadable file? 
Now that it has become so short.... just printing it out should be enough - and if the issue with the warning count mentioned in my comment above is dealt with.. then the suppression of the error status code will also become unnecessary

kfischer-okarin reviewed (commented):
Looks mostly good to me... added one little comment...

About the warnings.... doesn't slither have a `.slitherignore` file or something? Or can ignore files/issues via comments like other linters? 
I think doing checks via a max number of expected warnings is not a very safe/predictable way to do this... If someone removes `Base64.sol` without remembering to adjust these tests it will allow 5 totally different warnings to go unnoticed.....

haltakov committed:
Simplify the slither workflow

haltakov committed:
Work around remaining slither issues

haltakov reviewed (approved):
This looks good now. I did some tests and a couple of small fixes to simplify the `slither` workflow. There was one issue I wasn't able to solve, but I worked around it for now. Will create a GitHub issue for the real solution
